### ISSUE TITLE
Performance of flask-limiter with redis

### ISSUE THREAD

**Issue by lucas03** [alisaifee/flask-limiter#282]
We have used flask-limiter on our flask based app, but we are seeing some performance issues. We have implemented it with redis db like this:
```
def get_user_limit(username: Callable) -> str:
    """Get request user limit from config or use default limit."""
    username = username()
    return settings.limiter.users_rate_limits.get(username, settings.limiter.default_rate_limit)


def configure_limiter(application: FlaskApp) -> None:
    """Configure application."""
    limiter.init_app(application.app)


limiter: Final[Limiter] = Limiter(
    application_limits=[partial(get_user_limit, get_username)],  # shared limit for all routes
    key_func=get_username,
    storage_uri=settings.ratelimit_storage_url,
)
```
and it seemed to work well. Redis is a GCP memorystore version, instance should be enough (both CPU and memory wise). But we are spending a lot of time on `EVALSHA` command, from 4-35% of whole time spend by app.
<img src="https://monosnap.com/image/VBxtMCS9RensdNnRsOCPPVz4vbaSjV"/>
Which is of course too much :) our app is handling 4k requests per minute, and I would expect rate limiting on flask to be a bit more lightweight.

And if we zoom in on some requests, we can see how much time it takes per request:
<img src="https://monosnap.com/image/LoC9etcBuzldfYt2ovGVbZUPdp3vlr"/>
and the command that takes 726ms looks like this:
`EVALSHA 0d23cf1ccf81ad1452ccf307e1938d99997a6073 1 LIMITER/remote_user/global/1000/1/minute 60`
Is there a simple way to optimize this?

**Comment by alisaifee** [comment]
700ms or even any multiple of 100ms sounds horrible and should definitely not be the case. 

I made some assumptions based on your snippet - that you're using the fixed window strategy and have a global limit of 1000 requests/minute/user regardless of the endpoint.

To try and emulate this, I spun up a small gcp redis memorystore (16gb) and a tiny debian vm (e2 micro) and ran a toy flask app with one route that does literally nothing other than get ratelimited (1000/minute, and I randomly distributed "user" in 250 unique ips (faked). I ran the app on the single instance with 5 workers under uwsgi and hit it from outside the private network with 50 concurrent workers for a total of 500 repetitions:

```bash
❯ siege http://34.121.22.215 -r 500 -c 50
** SIEGE 4.0.4
** Preparing 50 concurrent users for battle.
The server is now under siege...
Transactions:                  25000 hits
Availability:                 100.00 %
Elapsed time:                 359.06 secs
Data transferred:               0.05 MB
Response time:                  0.71 secs
Transaction rate:              69.63 trans/sec
Throughput:                     0.00 MB/sec
Concurrency:                   49.74
Successful transactions:       25000
Failed transactions:               0
Longest transaction:            2.03
Shortest transaction:           0.12
 
```

Here's a snapshot of some newrelic transaction breakdown for the duration of the test:
![image](https://user-images.githubusercontent.com/79842/146271999-e3da5ed6-44dd-47fd-bc90-5e2ba3b916ce.png)

And this is the associated view of the redis instance:
![image](https://user-images.githubusercontent.com/79842/146272376-27980fd8-825a-4c5c-86db-ef1de34ac94e.png)

Which is what I would mostly expect.

For good measure I ran the same test with the moving window strategy (which is actually a bit more involved in terms of the lua script) and that too yielded similar throughput and time spent in redis.
